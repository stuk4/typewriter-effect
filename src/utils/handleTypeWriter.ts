

export const handleTypeWriter = (element:Element | null,text:string,interval:number = 20) => {
   
    if (!element) return;
      let charIndex = 0;
      const handleAddChar = ( ) => {
        element.classList.add('typewriter',"typewriter-blink");
        element.innerHTML += text[charIndex] === "\n" ? "<br>" : text[charIndex];
        charIndex++;
        if (charIndex < text.length) {
          setTimeout(handleAddChar, interval);
        }else{
          element.classList.remove("typewriter-blink");
        }
      }
      handleAddChar();
    
  }
