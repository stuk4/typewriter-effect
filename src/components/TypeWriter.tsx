import {createElement, useEffect, useRef} from 'react'
import { ITypeWriterProps } from '../interfaces/ITypeWriter';
import {handleTypeWriter} from '../utils/handleTypeWriter';

export const TypeWriter = ({
    elementType: Element,
    text,
    delay = 0,
    interval = 20,
    classNames
} : ITypeWriterProps) => {
    const refElement = useRef <HTMLElement | null > (null)
    useEffect(() => {
        const timeout = setTimeout(() => {
            handleTypeWriter(refElement.current, text, interval)
        }, delay)
        return() => {
            clearTimeout(timeout)
        }
    }, [])
    return createElement(Element, {ref: refElement,className:`${classNames ? classNames : ""}`,"data-testid":"typewriter-element"})
}
