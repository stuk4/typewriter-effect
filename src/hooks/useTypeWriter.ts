import { useCallback, useEffect } from "react";
import { handleTypeWriter } from "../utils/handleTypeWriter";

interface useTypeWriterProps{
  element:Element | null,
  interval?:number,
  text:string,
  delay?:number
}
export const useTypeWriter = ({element,interval=20,text,delay=0}:useTypeWriterProps) => {
  
  const handleTypeWriter = useCallback(() => {
    console.log("salto",element)
    if (!element) return;
      let charIndex = 0;
      const handleAddChar = ( ) => {
        element.classList.add('typewriter');
        element.innerHTML += text[charIndex] === "\n" ? "<br>" : text[charIndex];
        charIndex++;
        if (charIndex < text.length) {
          setTimeout(handleAddChar, interval);
        }
      }
      handleAddChar();
  },[element])

  useEffect(() => {
    const timeout = setTimeout(() =>{
        handleTypeWriter()
    },delay)
    return ( ) =>{
      clearTimeout(timeout)
    }
  }, [element])
 
}
