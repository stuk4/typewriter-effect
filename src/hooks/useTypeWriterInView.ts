import { useEffect } from "react";
import { handleTypeWriter } from "../utils/handleTypeWriter";

interface useTypeWriterProps{
  element:Element | null,
  interval?:number,
  text:string,
  delay?:number,
  inView:boolean,
}
export const useTypeWriter = ({element,interval=20,text,delay=0,inView}:useTypeWriterProps) => {
  
    useEffect(() => {
      if (inView) {
        setTimeout(() =>{
            handleTypeWriter(element, text,interval)
        },delay)
      }
  
  }, [inView])

 
}
