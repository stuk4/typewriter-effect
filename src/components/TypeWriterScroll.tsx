import React, { createElement } from 'react'
import { useInView } from 'react-intersection-observer'
import { useTypeWriter } from '../hooks/useTypeWriterInView'
import { ITypeWriterProps } from '../interfaces/ITypeWriter'

export const TypeWriterInScroll = ({
  elementType: Element,
  text,
  delay = 0,
  interval = 20,
  classNames

}:ITypeWriterProps) => {
  const [ref, inView,node] = useInView({
      rootMargin: '-100px 0px',
      triggerOnce: true

    })
  useTypeWriter({
      element:node?.target as Element,
      text:text,
      inView:inView,
      delay:delay,
      interval

    })
  return  createElement(Element, {ref: ref,className:`${classNames ? classNames : ""}`,"data-testid":"typewriter-element"})
}
