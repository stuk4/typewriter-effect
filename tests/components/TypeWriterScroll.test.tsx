import React from "react";
import {act, render, renderHook, waitFor} from '@testing-library/react';
import {TypeWriter} from '../../src/components/TypeWriter';

import {TypeWriterInScroll} from '../../src/components/TypeWriterScroll';
import {setupIntersectionMocking, resetIntersectionMocking} from 'react-intersection-observer/test-utils';
import {useTypeWriter} from "../../src/hooks/useTypeWriter";
import {handleTypeWriter} from "../../src/utils/handleTypeWriter";

/**
 * @jest-environment jsdom
*/

const ReactTestRenderer = require('react-test-renderer');


jest.mock('../../src/utils/handleTypeWriter');


describe('TypeWriter component', () => {

    beforeEach(() => {
        setupIntersectionMocking(jest.fn);
    });

    afterEach(() => {
        jest.clearAllMocks();
        resetIntersectionMocking();
    });

    it('renders the correct element type', () => {
        const tag = "h1"
        const {getByTestId} = render (<TypeWriterInScroll elementType={tag}
            text="Test text"/>);
        const element = getByTestId('typewriter-element');
        expect(element.tagName.toLowerCase()).toBe(tag);
    });

    it('should call handleTypeWriter when inView is true', async () => {
        const text = 'Hello, World!'
        const interval = 50;
        renderHook(() => useTypeWriter({
            element: document.createElement('div'),
            text:text ,
            inView: true,
            delay: 100,
            interval: interval
        }));

        await waitFor(() => {
            expect(handleTypeWriter).toHaveBeenCalledWith(document.createElement('div'), text, interval);
        })
    });

    it('should not call handleTypeWriter when inView is false', async () => {
        const text = 'Hello, World!'
        const interval = 50;
        renderHook(() => useTypeWriter({
            element: document.createElement('div'),
            text:text ,
            inView: false,
            delay: 100,
            interval: interval
        }));

        await waitFor(() => {
            expect(handleTypeWriter).not.toHaveBeenCalledWith();
        })
    });
    it('has the correct class name', () => {
        const className = 'test-class';
        const {getByTestId} = render (<TypeWriterInScroll elementType="h1" text="Test text"
            classNames={className}/>);
        const element = getByTestId('typewriter-element');
        expect(element.classList.contains(className)).toBe(true);

    });


    it('matches the snapshot', () => {
        const tree = ReactTestRenderer.create (<TypeWriterInScroll elementType="h1" text="Test text"/>).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
