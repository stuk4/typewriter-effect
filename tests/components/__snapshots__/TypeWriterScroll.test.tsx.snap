// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeWriter component matches the snapshot 1`] = `
<h1
  className=""
  data-testid="typewriter-element"
/>
`;
